patterns/behavioral/command.py

```diff
@@ -1,64 +1,131 @@
+method for each menu item and pass to it.
+
+*About the example
+We have a menu containing two items. Each item accepts a file name, one hides the file
+and the other deletes it. Both items have an undo option.
+Each item is a MenuItem class that accepts the corresponding command as input and executes
+it's `execute` method when it is pressed.
+
 *TL;DR
-Encapsulates all information needed to perform an action or trigger an event.
+Object oriented implementation of callback functions.
 
 *Examples in Python ecosystem:
 Django HttpRequest (without `execute` method):
- https://docs.djangoproject.com/en/2.1/ref/request-response/#httprequest-objects
+https://docs.djangoproject.com/en/2.1/ref/request-response/#httprequest-objects
 """
 
 import os
 
 
-class MoveFileCommand:
-    def __init__(self, src, dest):
-        self.src = src
-        self.dest = dest
+class HideFileCommand:
+    """
+    A command to hide a file given its name
+    """
+
+    def __init__(self):
+        # an array of files hidden, to undo them as needed
+        self._hidden_files = []
+
+    def execute(self, filename):
+        if os.path.isfile(filename):
+            print(f'hiding {filename}')
+
+            os.rename(filename, f'.{filename}')
+            self._hidden_files.append(filename)
+        else:
+            print(f'{filename} dose not exists to hide')
+
+    def undo(self):
+        if len(self._hidden_files) > 0:
+            filename = self._hidden_files.pop()
+
+            print(f'un-hiding {filename}')
+
+            os.rename(f'.{filename}', filename)
+
+
+class DeleteFileCommand:
+    """
+    A command to delete a file given its name
+    """
+    _deleted_files_path = 'trash'
+
+    def __init__(self):
+        # an array of deleted files, to undo them as needed
+        self._deleted_files = []
+
+        # create a directory to store deleted files
+        if not os.path.exists(self._deleted_files_path):
+            os.makedirs(self._deleted_files_path)
 
-    def execute(self):
-        self.rename(self.src, self.dest)
+    def execute(self, filename):
+        if os.path.isfile(filename):
+            print(f'deleting {filename}')
+
+            os.rename(filename, f'{self._deleted_files_path}/{filename}')
+            self._deleted_files.append(filename)
+        else:
+            print(f'{filename} dose not exists to delete')
 
     def undo(self):
-        self.rename(self.dest, self.src)
+        if len(self._deleted_files) > 0:
+            filename = self._deleted_files.pop()
+
+            print(f'un-deleting {filename}')
+
+            os.rename(f'{self._deleted_files_path}/{filename}', filename)
+
+
+class MenuItem:
+    """
+    The invoker class. Here it is items in a menu.
+    """
+
+    def __init__(self, command):
+        self._command = command
 
-    def rename(self, src, dest):
-        print("renaming {} to {}".format(src, dest))
-        os.rename(src, dest)
+    def on_do_press(self, filename):
```

Comment:
I think names like `on_do_press` and `on_undo_press` are harder to read then `rename`, `delete` or whatever else

Reply:
You are right, but the invoker (here is the menu item) does not know what is going to do.

Am I right?